javaslang/src/main/java/javaslang/collection/Iterator.java

```diff
@@ -467,51 +470,79 @@ public Short getNext() {
-        } else if (step * (from - toExclusive) >= 0) {
+        } else if (areEqual(from, toExclusive) || step.signum() == from.subtract(toExclusive).signum()) {
             return empty();
         } else {
-            return new AbstractIterator<Double>() {
+            if (step.compareTo(BigDecimal.ZERO) > 0) {
+                return new AbstractIterator<Double>() {
+                    BigDecimal i = from;
 
-                double prev = Double.NaN;
-                double curr = from;
-                boolean hasNext = true;
+                    @Override
+                    public boolean hasNext() {
+                        return i.compareTo(toExclusive) < 0;
+                    }
 
-                @Override
-                public boolean hasNext() {
-                    return hasNext;
-                }
+                    @Override
+                    public Double getNext() {
+                        final BigDecimal next = this.i;
+                        this.i = next.add(step);
+                        return next.doubleValue();
+                    }
+                };
+            } else {
+                return new AbstractIterator<Double>() {
+                    BigDecimal i = from;
+
+                    @Override
+                    public boolean hasNext() {
+                        return i.compareTo(toExclusive) > 0;
+                    }
 
-                @Override
-                public Double getNext() {
-                    final double next = curr;
-                    if ((step > 0 && curr + step >= toExclusive) || (step < 0 && curr + step <= toExclusive)) {
-                        hasNext = false;
-                    } else {
-                        prev = curr;
-                        curr += step;
-                        if (curr == prev) {
-                            hasNext = false;
-                        }
+                    @Override
+                    public Double getNext() {
+                        final BigDecimal next = this.i;
+                        this.i = next.add(step);
+                        return next.doubleValue();
                     }
-                    return next;
-                }
-            };
+                };
+            }
+        }
+    }
+
+    // TODO these two need a place to crash for the night :/
+
+    /* scale-independent equality */
+    static boolean areEqual(BigDecimal from, BigDecimal toExclusive) {
+        return from.compareTo(toExclusive) == 0;
+    }
+
+    /* parse infinite values also */
+    static final BigDecimal infinityMultiplier = BigDecimal.valueOf(1.000000000000001); // decrease by as little as possible so that it doesn't have a double representation anymore -> infinite
+
+    static BigDecimal asDecimal(double number) {
+        if (number == NEGATIVE_INFINITY) {
+            final BigDecimal result = BigDecimal.valueOf(Math.nextUp(NEGATIVE_INFINITY));
+            return result.multiply(infinityMultiplier);
```

Comment:
The boundaries can quickly be calculated using binary search:

```java
// -- doubles

private BigDecimal binarySearch(BigDecimal finite, BigDecimal infinite) {
    while (true) {
        final BigDecimal magicValue = finite.add(infinite).divide(BigDecimal.valueOf(2), HALF_UP);
        if (Double.isInfinite(magicValue.doubleValue())) {
            if (magicValue.equals(infinite)) {
                return magicValue;
            }
            infinite = magicValue;
        } else {
            finite = magicValue;
        }
    }
}

@Test
public void infinityAndBeyond() {
    final BigDecimal finite = BigDecimal.valueOf(Double.MAX_VALUE);
    final BigDecimal infinite = BigDecimal.valueOf(2).pow(Double.MAX_EXPONENT + 1);

    final BigDecimal negative = binarySearch(finite.negate(), infinite.negate());
    assertThat(negative.add(BigDecimal.ONE).doubleValue()).isNotEqualTo(Double.NEGATIVE_INFINITY);
    assertThat(negative.doubleValue()).isEqualTo(Double.NEGATIVE_INFINITY);

    final BigDecimal positive = binarySearch(finite, infinite);
    assertThat(positive.subtract(BigDecimal.ONE).doubleValue()).isNotEqualTo(Double.POSITIVE_INFINITY);
    assertThat(positive.doubleValue()).isEqualTo(Double.POSITIVE_INFINITY);
}
```

we could calculate and store this value in a static block / field.
I'm still trying to figure out why `179769313486231580793728971405303415079934132710037826936173778980444968292764750946649017977587207096330286416692887910946555547851940402630657488671505820681908902000708383676273854845817711531764475730270069855571366959622842914819860834936475292719074168444365510704342711559699508093042880177904174497792` is so special :p